### ISSUE TITLE
Introduce configuration using Custom Resources instead of annotations

### ISSUE THREAD

**Issue by jannfis** [argoproj-labs/argocd-image-updater#235]
**Is your feature request related to a problem? Please describe.**

Currently, Argo CD Image Updater is configured in two ways:

* The images and applications to be updated are configured by applying annotation on Argo CD's `Application` resources. This is cumbersome (need to for each application which uses a certain image) and error prone (e.g. typos). 

* Additionally registry configuration is done using a YAML within a ConfigMap resource, e.g. `registries.conf`

**Describe the solution you'd like**

We want to make the configuration of Argo CD Image Updater - both for registries and applications - to be a strongly typed one, so that creating a configuration becomes less error prone. Also, we want to decouple the image configuration from a single `Application` resource, so that it becomes easier to say "update image `foobar` in all applications matching a giving pattern/project/label selector`.

For this, two new CRDs should be introduced:

* `argoproj.io/ImageUpdateConfig` for describing images to be updated and mapping them to `Applications`
* `argoproj.io/RegistryConfig` for configuring custom registries to be used by Image Updater

The APIs have yet to be defined and are currently work in progress. But they will eventually replace the current way of configuring Image Updater.

An initial version of the APIs/CRDs will be attached to this issue once they are in a reviewable state.

**Comment by diranged** [comment]
@jannfis Just  curious if you've made any progress on this front?

**Comment by aburan28** [comment]
Is anyone actively working on this? Happy to take this feature request on

**Comment by jannfis** [comment]
@aburan28 Thanks for your offer. Before writing any code, we should define & agree on an API though. This will be a dry run task. There's a couple of things in the backlog too, that play into this. Feel free to come up with a proposal for an API (probably as a [Discussion](https://github.com/argoproj-labs/argocd-image-updater/discussions)). I do believe this will not be an easy task, and there will be lots of discussions and back and forths around it.

**Comment by aburan28** [comment]
Absolutely, sounds good. I will come up with a proposal and start a discussion on this topic

**Comment by chengfang** [comment]
Current `master` branch and [releases](https://github.com/argoproj-labs/argocd-image-updater/tags) 1.x.x are now based on CRD, see docs https://argocd-image-updater.readthedocs.io/en/stable/

[master-annotation-based](https://github.com/argoproj-labs/argocd-image-updater/tree/master-annotation-based) branch and prior releases 0.x.x are for the legacy design based on application annotations. See docs for prior release https://argocd-image-updater.readthedocs.io/en/release-0.17/
